{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "170e1e3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from itertools import chain\n",
    "\n",
    "import msprime\n",
    "import pyslim\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import SVG"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fa4d25e",
   "metadata": {},
   "source": [
    "Load the `trees` file generated by SLiM and inspect the contents:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "487cd7b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "slim_ts = pyslim.load(\"results/europe_tree_seq.trees\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c83a8203",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\"src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.readthedocs.io/en/latest/python-api.html#the-treesequence-class\"> Tree Sequence\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>530278</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>100000000.0</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>21044</td></tr>\n",
       "                      <tr><td>Total Size</td><td>81.3 MiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\"></span>\n",
       "                  <details open>\n",
       "                    <summary>dict</summary>\n",
       "                    \n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">SLiM:</span>\n",
       "                  <details >\n",
       "                    <summary>dict</summary>\n",
       "                    file_version: 0.6<br/>generation: 168500<br/>model_type: WF<br/>nucleotide_based: False<br/>separate_sexes: False<br/>spatial_dimensionality: xy<br/>spatial_periodicity: <br/>stage: late<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                </td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>2057175</td>\n",
       "                      <td>54.9 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>10522</td>\n",
       "                      <td>659.1 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>4 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>1.2 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>307490</td>\n",
       "                      <td>10.0 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>7</td>\n",
       "                      <td>3.2 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>1</td>\n",
       "                      <td>20.3 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>0</td>\n",
       "                      <td>8 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<pyslim.slim_tree_sequence.SlimTreeSequence at 0x7fca6c63aac0>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slim_ts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd22da9e",
   "metadata": {},
   "source": [
    "Check that all trees in the sequence are coalesced:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f23db6ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_coalescence(ts):\n",
    "    \"\"\"Check whether all trees in a sequence completely coalesced.\"\"\"\n",
    "    for t in ts.trees():\n",
    "         assert t.num_roots == 1, (\"not coalesced! on segment {} to {}\".\n",
    "              format(t.interval[0], t.interval[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1aac3d07",
   "metadata": {},
   "outputs": [],
   "source": [
    "check_coalescence(slim_ts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37e5c051",
   "metadata": {},
   "source": [
    "## Simplification\n",
    "\n",
    "Realistically, we will never have all data that we simulated, which is why we subsample to _N_ chromosomes from each population and then simplify the tree sequence coalescent structure accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9acce5ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subsample to 1 chromosome from each population\n",
    "k = 1\n",
    "samples = []\n",
    "\n",
    "for pop in slim_ts.populations():\n",
    "    pop_sample = random.sample(list(slim_ts.samples(pop.id)), k=k)\n",
    "    samples.append(pop_sample)\n",
    "\n",
    "# flatten the lists\n",
    "samples = list(chain.from_iterable(samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0d33182a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts, mapping = slim_ts.simplify(samples=samples, map_nodes=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54eac6c8",
   "metadata": {},
   "source": [
    "## Plot the first tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9d2b2a61",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = ts.first()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "239ddb49",
   "metadata": {},
   "source": [
    "Put together a table of population ➡ color assignments for coloring the nodes of the tree:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "12ce6a76",
   "metadata": {},
   "outputs": [],
   "source": [
    "color_map = {0: \"red\", 1: \"blue\", 2: \"green\", 3: \"brown\", 4: \"orange\", 5: \"purple\", 6: \"black\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "47811064",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_names = open(\"model-europe/names.txt\").read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bdec7c1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>pop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>red</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>blue</td>\n",
       "      <td>OOA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>green</td>\n",
       "      <td>EHG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>brown</td>\n",
       "      <td>ANA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>orange</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>purple</td>\n",
       "      <td>YAM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>black</td>\n",
       "      <td>YAM_migr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        0       pop\n",
       "0     red       AFR\n",
       "1    blue       OOA\n",
       "2   green       EHG\n",
       "3   brown       ANA\n",
       "4  orange       EUR\n",
       "5  purple       YAM\n",
       "6   black  YAM_migr"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop_colors = pd.DataFrame.from_dict(color_map, orient=\"index\")\n",
    "pop_colors[\"pop\"] = pop_names\n",
    "pop_colors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d739bdf",
   "metadata": {},
   "source": [
    "Assign color to each node:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1d3bf1db",
   "metadata": {},
   "outputs": [],
   "source": [
    "node_colors = {i: color_map[tree.population(i)] for i in tree.nodes()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b2ac2d99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"400\" version=\"1.1\" width=\"500\"><defs><style type=\"text/css\"><![CDATA[.tree-sequence .background path {fill: #808080; fill-opacity:0}.tree-sequence .background path:nth-child(odd) {fill-opacity:.1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold}.axes, .tree {font-size: 14px; text-anchor:middle}.y-axis line.grid {stroke: #FAFAFA}.y-axis > .lab text {transform: translateX(0.8em) rotate(-90deg)}.x-axis .tick g {transform: translateY(0.9em)}.x-axis > .lab text {transform: translateY(-0.8em)}.axes line, .edge {stroke:black; fill:none}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree t0\"><g class=\"plotbox\"><g class=\"node n9212 p0 root\" transform=\"translate(268.482 26.8)\"><g class=\"a9212 node n2490 p0\" transform=\"translate(145.804 242.209)\"><g class=\"a2490 i3 leaf node n3 p3 sample\" transform=\"translate(-32.8571 72.2955)\"><path class=\"edge\" d=\"M 0 0 V -72.2955 H 32.8571\"/><rect class=\"sym\" height=\"6\" style=\"fill:brown;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a2490 i5 leaf node n5 p5 sample\" transform=\"translate(32.8571 94.0422)\"><path class=\"edge\" d=\"M 0 0 V -94.0422 H -32.8571\"/><rect class=\"sym\" height=\"6\" style=\"fill:purple;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -242.209 H -145.804\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">2490</text></g><g class=\"a9212 node n6075 p0\" transform=\"translate(-145.804 132.51)\"><g class=\"a6075 i0 leaf node n0 p0 sample\" transform=\"translate(-69.8214 213.89)\"><path class=\"edge\" d=\"M 0 0 V -213.89 H 69.8214\"/><rect class=\"sym\" height=\"6\" style=\"fill:red;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6075 node n3766 p0\" transform=\"translate(69.8214 73.0204)\"><g class=\"a3766 i2 leaf node n1 p1 sample\" transform=\"translate(-73.9286 69.3477)\"><path class=\"edge\" d=\"M 0 0 V -69.3477 H 73.9286\"/><rect class=\"sym\" height=\"6\" style=\"fill:blue;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a3766 node n1084 p1\" transform=\"translate(73.9286 73.6487)\"><g class=\"a1084 i1 leaf node n4 p4 sample\" transform=\"translate(49.2857 67.2213)\"><path class=\"edge\" d=\"M 0 0 V -67.2213 H -49.2857\"/><rect class=\"sym\" height=\"6\" style=\"fill:orange;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a1084 node n321 p1\" transform=\"translate(-49.2857 19.427)\"><g class=\"a321 i4 leaf node n2 p2 sample\" transform=\"translate(-32.8571 22.6648)\"><path class=\"edge\" d=\"M 0 0 V -22.6648 H 32.8571\"/><rect class=\"sym\" height=\"6\" style=\"fill:green;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a321 i6 leaf node n6 p6 sample\" transform=\"translate(32.8571 40.0622)\"><path class=\"edge\" d=\"M 0 0 V -40.0622 H -32.8571\"/><rect class=\"sym\" height=\"6\" style=\"fill:black;\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -19.427 H 49.2857\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">321</text></g><path class=\"edge\" d=\"M 0 0 V -73.6487 H -73.9286\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:blue;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">1084</text></g><path class=\"edge\" d=\"M 0 0 V -73.0204 H -69.8214\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">3766</text></g><path class=\"edge\" d=\"M 0 0 V -132.51 H 145.804\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6075</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" style=\"fill:red;\"/><text class=\"lab\" transform=\"translate(0 -11)\">9212</text></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SVG(tree.draw(width=500, height=400, node_colours=node_colors))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1299c2d5",
   "metadata": {},
   "source": [
    "## Add mutations to the trees\n",
    "\n",
    "I just ran [this](https://pyslim.readthedocs.io/en/latest/tutorial.html#adding-neutral-mutations-to-a-slim-simulation) but did not investigate further."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "634bf0be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The tree sequence now has 46738 mutations\n"
     ]
    }
   ],
   "source": [
    "ts = pyslim.SlimTreeSequence(msprime.mutate(ts, rate=1e-8, keep=True))\n",
    "\n",
    "print(f\"The tree sequence now has {ts.num_mutations} mutations\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6982da34",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes = ts.genotype_matrix()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "25d649c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46738, 7)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d2b7d30",
   "metadata": {},
   "source": [
    "We generated genotypes of 7 chromosomes as we wanted..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e04ea45c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0, 1, 0, ..., 0, 0, 0],\n",
       "       [0, 0, 0, ..., 0, 1, 0],\n",
       "       [0, 0, 0, ..., 1, 0, 0],\n",
       "       ...,\n",
       "       [0, 1, 1, ..., 1, 1, 1],\n",
       "       [0, 0, 0, ..., 0, 0, 0],\n",
       "       [1, 0, 0, ..., 0, 0, 0]], dtype=int8)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4da91712",
   "metadata": {},
   "source": [
    "## Locations of owners of those sampled chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e504b5ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7, 3)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts.individual_locations.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d733b7cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[3.33997965e+002, 4.40256422e+001, 0.00000000e+000],\n",
       "       [2.33445346e+002, 2.94061084e+002, 1.42173718e-311],\n",
       "       [6.85491679e+002, 1.92478123e+002, 7.72521670e+007],\n",
       "       [4.71898318e+002, 2.15655492e+002, 1.00000000e+000],\n",
       "       [4.17080406e+002, 4.43362408e+002, 7.17234590e-312],\n",
       "       [4.90827829e+002, 3.47596402e+002, 6.01814120e+007],\n",
       "       [4.31505486e+002, 3.25657641e+002, 6.91469389e+001]])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts.individual_locations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b588884b",
   "metadata": {},
   "source": [
    "So each individual has a location attached to it in three-dimensions, but the third one is \"basically zero\" (which makes sense, because we run the SLiM simulation on a two-dimensional map)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8446a96",
   "metadata": {},
   "source": [
    "The metadata of each individual (most importantly the time and location) can be accessed via tskit/pyslim API:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a36ba24b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': 0, 'flags': 65536, 'location': array([333.99796546,  44.02564225,   0.        ]), 'parents': array([], dtype=int32), 'nodes': array([0], dtype=int32), 'population': 0, 'time': 0.0, 'metadata': {'pedigree_id': 873019302, 'age': -1, 'subpopulation': 0, 'sex': -1, 'flags': 0}}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts.individual(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98578eaf",
   "metadata": {},
   "source": [
    "The simplest thing to start with would be generating genotype data with msprime (as shown above and in the documentation) and then extract the true known locations of individuals using the tskit functionality."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
